counsel.el

```diff
@@ -1274,21 +1274,20 @@ Typical value: '(recenter)."
       (prog1
           (setq ivy--old-cands
                 (cl-remove-if-not
-                 (lambda (x)
-                   (ignore-errors
-                     (when (string-match "^[^:]+:[^:]+:" x)
-                       (setq x (substring x (match-end 0)))
-                       (if (stringp regexp)
-                           (string-match regexp x)
-                         (let ((res t))
-                           (dolist (re regexp)
-                             (setq res
-                                   (and res
-                                        (ignore-errors
-                                          (if (cdr re)
-                                              (string-match (car re) x)
-                                            (not (string-match (car re) x)))))))
-                           res)))))
+                 (let* ((grep-loc-re "^[^:]+:[^:]+:")
+                        (re (and (stringp regexp)
+                                 (concat grep-loc-re regexp))))
+                   (if re (lambda (x) (string-match re x))
+                     (lambda (x)
+                       (ignore-errors
+                         (when (string-match grep-loc-re x)
+                           (let ((start (match-end 0)))
+                             (cl-every
+                              (lambda (re)
+                                (if (cdr re)
+                                    (string-match (car re) x start)
+                                  (not (string-match (car re) x start))))
```

Comment:
Ditto re: `string-match-p`.

I would write this condition as
```el
(eq (not (string-match-p (car re) x start))
    (not (cdr re)))
```
but it's up to you. (At the very least I would write `(if (string-match-p ...) (cdr re) (not (cdr re)))` so as to avoid repeating the more complicated logic.)

Reply:
I did not really take great care of the second case before knowing if it is actually used. But I agree with your comment.